rust/src/http2/detect.rs

```diff
@@ -0,0 +1,512 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use super::http2::{HTTP2FrameTypeData, HTTP2Transaction};
+use super::parser;
+use crate::core::{STREAM_TOCLIENT, STREAM_TOSERVER};
+use std::ffi::CStr;
+use std::mem::transmute;
+use std::str::FromStr;
+
+#[no_mangle]
+pub extern "C" fn rs_http2_tx_has_frametype(
+    tx: *mut std::os::raw::c_void, direction: u8, value: u8,
+) -> std::os::raw::c_int {
+    let tx = cast_pointer!(tx, HTTP2Transaction);
+    if direction & STREAM_TOSERVER != 0 {
+        for i in 0..tx.frames_ts.len() {
+            if tx.frames_ts[i].header.ftype as u8 == value {
+                return 1;
+            }
+        }
+    }
+    if direction & STREAM_TOCLIENT != 0 {
+        for i in 0..tx.frames_tc.len() {
+            if tx.frames_tc[i].header.ftype as u8 == value {
+                return 1;
+            }
+        }
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_http2_parse_frametype(
+    str: *const std::os::raw::c_char,
+) -> std::os::raw::c_int {
+    let ft_name: &CStr = CStr::from_ptr(str); //unsafe
+    if let Ok(s) = ft_name.to_str() {
+        if let Ok(x) = parser::HTTP2FrameType::from_str(s) {
+            return x as i32;
+        }
+    }
+    return -1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_http2_tx_has_errorcode(
+    tx: *mut std::os::raw::c_void, direction: u8, code: u32,
+) -> std::os::raw::c_int {
+    let tx = cast_pointer!(tx, HTTP2Transaction);
+    if direction & STREAM_TOSERVER != 0 {
+        for i in 0..tx.frames_ts.len() {
+            match tx.frames_ts[i].data {
+                HTTP2FrameTypeData::GOAWAY(goaway) => {
+                    if goaway.errorcode == code {
+                        return 1;
+                    }
+                }
+                HTTP2FrameTypeData::RSTSTREAM(rst) => {
+                    if rst.errorcode == code {
+                        return 1;
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+    if direction & STREAM_TOCLIENT != 0 {
```

Comment:
Hmm...
This is interesting.
I thought `direction` was referring to the rule.
And so it could be both `STREAM_TOCLIENT` and `STREAM_TOSERVER`
So, in the case of bidirectional signatures, this will be called twice, right ?